Add VMI backend storage and persistent TPM support

jean-edouard opened the pull request:
**What this PR does / why we need it**:
This PR contains a first commit that implements a way to persist VM files on the backend side.
The second commit then uses that to add persistent TPM support.
The third commit modifies the behavior of the backend storage PVC to add support for migration.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This is better reviewed commit by commit. I can split into 2 PRs but could only open £2 after £1 is merged.
I mostly followed the design proposal: https://github.com/kubevirt/community/pull/170

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
TPM VM device can now be set to persistent
```

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-compute

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.24-sig-storage

vladikr commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2170,6 +2170,10 @@ type KubeVirtConfiguration struct {
+
+	// BackendStorageClass is the name of the storage class to use for the PVCs created to preserve VM state, like TPM.
+	// The storage class must support RWX in filesystem mode.
+	BackendStorageClass string `json:"backendStorageClass,omitempty"`
```
To me `BackendStorageClass` sounds too generic and needs some kind of a prefix to explain what it's meant for.

jean-edouard replied:
I agree. I've had a hard time naming this thing, suggestions are welcome!

vasiliy-ul replied:
Maybe just `VMStateStorageClass`?

jean-edouard replied:
Done

vladikr commented on pkg/virt-controller/watch/migration.go:
```diff
@@ -843,7 +846,11 @@ func (c *MigrationController) handleTargetPodCreation(key string, migration *vir
 			}
 		}
-		return c.createTargetPod(migration, vmi, sourcePod)
+		backendStoragePVC, err := backendstorage.CreateIfNeeded(vmi, c.clusterConfig, c.clientset, backendstorage.IsMigration)
```
I'm a bit lost. Could you please explain why do we need to do this?
I mean why can't we simply skip this volume during migration and re-use the original on the destination? ( same as we do with other volumes)

jean-edouard replied:
That would be nice but I don't think we can.
When talking about volumes used by the VM it's not a problem, since the VM only ever runs in one place.
However, in this case, the contents of the PVC would be in use in 2 places at once (source container and target container).
In the case of swtpm, we could end up with corrupt data. In fact, swtpm takes a file-based lock to prevent exactly that, so the migration actually fails when using the original in the destination.

vladikr replied:
> However, in this case, the contents of the PVC would be in use in 2 places at once (source container and target container).

How can this be in use in two places? During the migration, the target domain is paused and will only be unpaused when the source is dead.
Why would this case be any different?

> In the case of swtpm, we could end up with corrupt data. In fact, swtpm takes a file-based lock to prevent exactly that, so the migration actually fails when using the original in the destination.
We would end up in this situation for any migration, but this is not happening because of what I've described above. Also, source qemu normally locks so the destination won't write into it accidentally.

jean-edouard replied:
> > However, in this case, the contents of the PVC would be in use in 2 places at once (source container and target container).
> 
> How can this be in use in two places? During the migration, the target domain is paused and will only be unpaused when the source is dead. Why would this case be any different?

Yes, the target domain is paused but not the target pod! This case is different because the contents of the PVC are actually accessed by the pod itself, not the domain.

> 
> > In the case of swtpm, we could end up with corrupt data. In fact, swtpm takes a file-based lock to prevent exactly that, so the migration actually fails when using the original in the destination.
> 
> We would end up in this situation for any migration, but this is not happening because of what I've described above. Also, source qemu normally locks so the destination won't write into it accidentally.

Yup, this again applies to the domain but not to its hosting pod.

vladikr replied:
> > > However, in this case, the contents of the PVC would be in use in 2 places at once (source container and target container).
> > 
> > 
> > How can this be in use in two places? During the migration, the target domain is paused and will only be unpaused when the source is dead. Why would this case be any different?
> 
> Yes, the target domain is paused but not the target pod! This case is different because the contents of the PVC are actually accessed by the pod itself, not the domain.

Can you please explain why the pod is accessing this content (especially if the domain is not running) if this content belongs to the domain?

Filesystem volumes that we attach to the VMIs should behave similarly. These volumes are mounted in the pod (compute container) and have a diskX.img file that is being used by qemu. 
It seems to me like a similar scenario. If not, can you please explain why not?

jean-edouard replied:
Ok I see your point now... Even when coming from the domain, most I/O operations ultimately happen in the backend. I guess a more accurate answer to your question is that qemu is equipped to handle cases where the same volume is mounted on both the source and the target, and swtpm is not... We could open an issue against swtpm and switch to using just 1 PVC if/when the issue gets resolved.

vladikr replied:
> Ok I see your point now... Even when coming from the domain, most I/O operations ultimately happen in the backend. I guess a more accurate answer to your question is that qemu is equipped to handle cases where the same volume is mounted on both the source and the target, and swtpm is not... We could open an issue against swtpm and switch to using just 1 PVC if/when the issue gets resolved.

Qemu is able to handle this mainly because the domain is paused and will not be unpaused until the source is destroyed. The same should happen with swptm. What is the reason this service would generate any I/O if qemu, which is its only consumer, is paused?

jean-edouard replied:
Yes, great question. It almost feels like that lock swtpm has in place is there to block this very scenario, I think we should ask about it, I'll open an issue.

vladikr replied:
I'm not sure which lock would that be, but if this code has been accepted into libvirt then Im sure this volume should be treated as any other. 

Honestly, I'd be against us developing an additional mechanism to handle PVCs , instead this volume should be treated the same way as any other filesystem volume we attach to the guests.

jean-edouard replied:
This lock (inside virt-launcher/compute, running as non-root): /var/run/kubevirt-private/libvirt/qemu/swtpm/09d086e6-ba5d-4647-8f7d-c16d17ae20c0/tpm2/.lock

Please keep in mind that we do not attach this volume to the guest! Libvirt is not aware of it, and the libvirt TPM option do not have a way to provide a specific disk or even location for storing the TPM state.

vladikr reviewed (commented):
Great work Jed. 
I have a few questions, mainly about live migration.

alicefr commented on pkg/virt-controller/services/rendervolumes.go:
```diff
@@ -281,6 +283,36 @@ func withAccessCredentials(accessCredentials []v1.AccessCredential) VolumeRender
+	return func(renderer *VolumeRenderer) error {
+		if backendstorage.HasPersistentTPMDevice(vmi) {
+			volumeName := vmi.Name + "-tpm"
+			pvcName := backendstorage.PVCPrefix + vmi.Name
```
Can we create a function that returns the name of the PVC for the tpm? So, you can also use it in [backend-storage](https://github.com/kubevirt/kubevirt/blob/079fc5b1c3846604252931bcdd359e3c62a746bd/pkg/backend-storage/backend-storage.go#L47)

jean-edouard replied:
Will do!

alicefr commented on tests/vmi_tpm_test.go:
```diff
@@ -83,3 +89,86 @@ var _ = Describe("[sig-compute]vTPM", func() {
+		It("should persist TPM secrets across restarts", func() {
+			By("Setting nfs-csi as the backend storage class")
+			kv := util.GetCurrentKv(virtClient)
+			kv.Spec.Configuration.BackendStorageClass = "nfs-csi"
```
Don't we need to skip the test `nfs-csi` storage class isn't installed?

jean-edouard replied:
Indeed, will do.

jean-edouard replied:
I switched to using the proper libstorage function and added a skip.

alicefr commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -34,27 +36,37 @@ func isBackendStorageNeeded(vmi *corev1.VirtualMachineInstance) bool {
+		}
+	} else if len(pvcList.Items) > 1 {
+		return "", fmt.Errorf("found too many backend storage PVCs for %s", vmi.Name)
+	} else if len(pvcList.Items) == 1 {
```
You can put all in one confition:
```golang
if if len(pvcList.Items) == 1 && !isMigration
```

alicefr commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -65,11 +77,63 @@ func CreateIfNeeded(vmi *corev1.VirtualMachineInstance, clusterConfig *virtconfi
 			VolumeMode:       &modeFile,
 		},
 	}
+	// Whatever owns the VMI (i.e. a VM) now owns the PVC
```
When the VM obejct is deleted, do we need also to delete the PVC?

jean-edouard replied:
Yes. I'm not sure how this mechanism works, I'll look into it.

jean-edouard replied:
The PVC indeed gets deleted with the VM, I added a test for it.

alicefr replied:
ok, what happen if we simpy create a VMI? This doesn't have a owner reference. What happen to the PVC in this case?

jean-edouard replied:
In that case the PVC stays, since we don't know if/when that VMI will be started again...
The alternative is to prevent the creation of VMIs with persistent settings enabled.

vasiliy-ul replied:
So while the VMI is shut down, if I start another VMI with the same name it will pick the existing PVC with TPM state, right? I am not deeply familiar with TPM, but does that mean that from the newly created VMI I can in theory get access to all the stored secrets?

jean-edouard replied:
Potentially yes. Same thing if a VM if stopped and you create a VMI with its name.
However:
- If the libvirt UUID is different, the existing TPM will be wiped out and a blank one will be created
- If the original VMI took ownership of the TPM, the new VMI will need the administrator password to do anything on it
- If the secrets were sealed against a set of PCRs that are unique to the original VMI, the new one will not be able to see them
- Most importantly, this should be seen as a convenience feature rather than a security one. The whole premise of TPM is the hardware vault protecting its flash memory...

vasiliy-ul replied:
I see. Thanks for clarification. So:

> If the libvirt UUID is different, the existing TPM will be wiped out and a blank one will be created

This basically means that there is no point in allowing persistent TPM settings for VMIs since UUID will be different every time, am I right? 

And BTW, just curious: I never checked that, but if I create a VM, then each time that I start/stop it the UUID will be the same, right?

vasiliy-ul replied:
I thought a new UUID was regenerated each time. Interesting...

jean-edouard replied:
Indeed, the libvirt docs says a random UUID is generated for each new domain (and every boot of a VM/VMI effectively creates a new domain).
However, I found this, though it is specific to VMs:
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/watch/vm.go#L1395

I just tried to verify my claim that VMIs keep getting the same UUID and actually got different values across reboots! Not sure what changed or maybe my initial observation was bogus...
However, even for VMIs we do request a specific UUID in the domain XML, so we have control over it. Good to know in case we want to ensure it's always the same value or something.

alicefr commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -65,11 +77,63 @@ func CreateIfNeeded(vmi *corev1.VirtualMachineInstance, clusterConfig *virtconfi
+}
 
-	_, err = client.CoreV1().PersistentVolumeClaims(vmi.Namespace).Create(context.Background(), pvc, metav1.CreateOptions{})
+func RemoveOldPVCFor(vmi string, pod *v1.Pod, client kubecli.KubevirtClient) {
```
Which PVCs is this function supposed to remove? The backed PVC for the TPM?

jean-edouard replied:
The backend PVC used by the VMI pre-migration.

alicefr replied:
The maybe we should rename the function to something like `RemoveOldBackendPVCFor`

alicefr commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -65,11 +77,63 @@ func CreateIfNeeded(vmi *corev1.VirtualMachineInstance, clusterConfig *virtconfi
+		return
+	}
+	for _, pvc := range pvcList.Items {
+		if pvc.Name != toKeep {
```
What is the PVC that you want to keep? Are we deleting here also other PVCs? Like for example other RWX PVCs that the VMI might own?

jean-edouard replied:
We want to keep the PVC that was created for the migration target.
The `List()` call is restricted to the label selector I created, however I need to change its name (`kubevirt.io/vm=`) to something more obvious.

vasiliy-ul commented on tests/vmi_tpm_test.go:
```diff
@@ -83,3 +89,97 @@ var _ = Describe("[sig-compute]vTPM", func() {
+	})
+
+	Context("with persistent TPM VM option enabled", func() {
+		It("[Serial]should persist TPM secrets across migrations and restarts", func() {
```
Would it be better to have two separate tests: one for `migrations` another one - for `restarts`? IMHO, debugging smaller tests is usually easier.

jean-edouard replied:
That's a good point, I actually went back-and-forth on the idea.
I chose to do all in 1 test because I think there's value in testing stop-start after migrating.
However, it would also be nice to test migrating after start-stop... Maybe I can try to create a table with 2 "operations", where the second one can be nil, and then test 4 entries:
- start-stop + nil
- migration + nil
- start-stop + migration
- migration + start-stop
That would take a while to run but it's probably good to have anyway.

jean-edouard replied:
Done, PTAL!

vasiliy-ul replied:
Looks good. Is it intentional that you skipped `start-stop + nil` and `migration + nil`?

jean-edouard replied:
Yeah. I can add the +nil cases but it feels redundant.

jean-edouard commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,65 @@
+	}
+	pvc := &v1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: PVCPrefix + vmi.Name,
```
We want it to, yes.
I set its owner to the same value(s) as the VMI's owner, not sure if that's enough, I need to add a test for it.

jean-edouard replied:
That wasn't the case anymore. I just pushed a commit to set the PVC owner to the VM object (or whatever the VMI owner is).
For discrete VMIs (i.e. not tied to a VM), I decided to set to owner of the PVC to the VMI, rendering it useless, but also avoiding garbage PVCs and security concerns.
PTAL!

jean-edouard commented:
/hold
Thanks a lot for all the reviews!!
Holding while I address PR comments and work on fixing the nonroot failure.

jean-edouard commented:
/hold cancel

jean-edouard commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,138 @@
+	toRemove := ""
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim != nil {
+			toRemove = volume.PersistentVolumeClaim.ClaimName
```
Hmm yeah good catch, I'm clearly missing something there, not sure what I was thinking :confused: 
Will update

jean-edouard replied:
Added a check on the name, PTAL!

vasiliy-ul replied:
Makes sense now. Thanks!

jean-edouard commented:
Added a commit to improve TPM directory creation in non-root. Used an initContainer instead of creating unnecessary emptydirs.

jean-edouard commented:
/hold
This PR contains a workaround because swtpm doesn't currently support shared storage migration.
However, a PR has been opened upstream https://github.com/stefanberger/swtpm/pull/732 to fix that, which is much better than this workaround.

rmohr commented:
Pretty great work. I have a similar requirement for persisting EFI_VARS files. I wonder if it would make sense to have one central PVC for all VMs and backup/restore through virt-handler. When thinking about scale, we would need quite some PVCs (hundreds)  and while one can request small PVC sizes, internally storage providers sometimes have (if I remember correctly) miniaml sized of 1Gi.

jean-edouard commented:
> Pretty great work. I have a similar requirement for persisting EFI_VARS files. I wonder if it would make sense to have one central PVC for all VMs and backup/restore through virt-handler. When thinking about scale, we would need quite some PVCs (hundreds) and while one can request small PVC sizes, internally storage providers sometimes have (if I remember correctly) miniaml sized of 1Gi.

Yes, the idea is indeed for that PVC to be multi-purpose, and EFI_VARS is the obvious other candidate.
The design proposal briefly touches on it:
https://github.com/kubevirt/community/pull/170/files#diff-02388d082a9acd98303b1858857a485ebd993e2518de159ebce66e342d7b2400R4

However, keeping per-VM PVCs is probably best, as discussed in the proposal PR.
I had the same belief about a minimum 1Gi PVC allocation size for some providers, but it doesn't really seems to be the case:
https://github.com/kubevirt/community/pull/170#discussion_r841873699
(actually, v1 of the design proposal suggested one central PVC for that reason, but I was convinced to change it)

In its current state, this PR includes a commit (https://github.com/kubevirt/kubevirt/pull/8156/commits/f9dacb9f1002f27b5a9af525ad546fafdce17063) to work around the fact that swtpm doesn't support shared storage between migration source and target.
This workaround makes a copy of the PVC on each migration, which works but was deemed too ugly to keep, and I am currently waiting for a proper upstream fix instead.
The upstream fix involves a PR in swtpm, which is merged (https://github.com/stefanberger/swtpm/pull/732) and a patchqueue against libvirt, which is not (https://listman.redhat.com/archives/libvir-list/2022-August/233769.html).

Since this PR doesn't really make sense without at least one use case, it is currently in limbo...
Maybe it would make sense to revive it, using EFI_VARS as the initial use-case instead? I do not currently have time to work on that myself, but I hopefully will in the future (or, even better, the upstream changes will land and we'll be able to do workaround-less TPM).
In any case, please feel free to take this over, make it work with EFI_VARS and PR it, I'd then just use that as a base for future vTPM storage.

tanaypatankar commented:
> The upstream fix involves a PR in swtpm, which is merged ([stefanberger/swtpm#732](https://github.com/stefanberger/swtpm/pull/732)) and a patchqueue against libvirt, which is not ([https://listman.redhat.com/archives/libvir-list/2022-August/233769.html](https://listman.redhat.com/archives/libvir-list/2022-August/233769.html)).

Both these PRs have been merged now @jean-edouard. I am looking into a use-case for persistent TPM. How do we proceed from here?

jean-edouard commented:
> > The upstream fix involves a PR in swtpm, which is merged ([stefanberger/swtpm#732](https://github.com/stefanberger/swtpm/pull/732)) and a patchqueue against libvirt, which is not (https://listman.redhat.com/archives/libvir-list/2022-August/233769.html).
> 
> Both these PRs have been merged now @jean-edouard. I am looking into a use-case for persistent TPM. How do we proceed from here?

Hi! Yes, upstream swtpm and libvirt should now support shared storage migration, which is needed to remove the ugly workaround from this PR.
swtpm also released v0.8.0, which contains the fix.
However, Centos 9 hasn't updated its swtpm package to v0.8.0, so we can't upgrade virt-launcher just yet.
I'm not sure what their timeline is for that, but I'll try to figure it out.
As soon as Centos 9 makes swtpm v0.8.0 available (as well as whichever libvirt version contains the corresponding fix), I'll PR a virt-launcher base OS bump, then update this PR by removing the migration workaround.

tanaypatankar commented:
libvirt v8.10.0 contains the fix for migration on shared storage.

rmohr commented:
/remove-lifecycle rotten

rmohr commented on pkg/container-disk/container-disk.go:
```diff
@@ -230,6 +230,18 @@ func GenerateKernelBootInitContainer(vmi *v1.VirtualMachineInstance, imageIDs ma
+	return &kubev1.Container{
+		Name:    "tpm-dir",
+		Image:   image,
+		Command: []string{"mkdir", "-p", "/var/run/kubevirt-private/libvirt/qemu/swtpm"},
```
Hm, init containers are in principle a little bit problematic and have for instance quite a bit of influence on the vm startup time.

Can virt-launcher just create the directory or virt-handler?

jean-edouard replied:
The problem with such old code is that I don't quite remember why I did things that way!
I do remember there was a good reason for this, might be SELinux-related, I wish I had been more comment-generous...
I'll test without it and keep you posted!

alicefr replied:
@jean-edouard can you please explain why you create this directory?

jean-edouard replied:
This is where the swtpm state file will be created. We want to mount the persistent storage there so the state will be automatically preserved, without having to resort to symlinks that SELinux won't like.

alicefr replied:
Sorry, it is still unclear to me. Isn't the state stored in the PVC? Isn't the mount and SELinux handled directly by k8s

jean-edouard replied:
`compute` will request to have PVC volume mounted under `/var/run/kubevirt-private/libvirt/qemu/swtpm`. I believe the directory needs to exist for the container to start successfully. 
Actually it's possible that the directory does get created by k8s if absent, but with the wrong permissions, in which case this might actually be something we can fix in virt-handler, as suggested by Roman. I will test that.

jean-edouard replied:
Yeah it's a permission thing... I just tested without the init container and got `panic: open /var/run/kubevirt-private/libvirt/qemu.conf: permission denied`
Is it a CRI bug? Does it really make sense to implement the logic for virt-handler to intervene before virt-launcher touches anything under `/var/run/kubevirt-private/libvirt/`?

vasiliy-ul replied:
Maybe adding `fsGroup: 107` to virt-launcher pod can solve the isue?

alicefr replied:
mh.. can you identify where the permissions are wrong? If it is on the directory or on the files? Do we set the fsGroup in virt-launcher, [here](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#configure-volume-permission-and-ownership-change-policy-for-pods) you can see in more details how the permission work

jean-edouard replied:
Ok so, without the init container, `/var/run/kubevirt-private/libvirt` is owned by `root` (bad) and has the SELinux label `container_file_t` (good). I believe `fsGroup` (which we don't set) would probably fix that (but I don't know if/how things like PSA restrict `fsGroup`).
However, `/var/run/kubevirt-private/libvirt/qemu` is not only owned by root, but also has the SELinux label `system_u:object_r:nfs_t:s0`, which is a bigger problem (I use the nfs-csi storage class for testing).

alicefr replied:
IIRC, nfs-csi isn't properly setting the ownership. @akalenyu do you recall? It might be a problem of that particular storage provider

akalenyu replied:
So no-provisioner NFS will not respect fsGroup at all, but nfs-csi should https://github.com/kubevirt/kubevirtci/blob/main/cluster-provision/k8s/1.26-centos9/manifests/nfs-csi/csi-nfs-driverinfo.yaml#L11

jean-edouard replied:
After setting fsGroup to 107, both directories are now owned by `root:qemu`, but `libvirt` has a mode of `drwxr-sr-x`, which doesn't allow qemu to create files under them!
`fsGroup` also doesn't fix the SELinux label issue (unsurprisingly).

alicefr replied:
Can you try to set the PVC explicitly not read-only? The SELinux seems to be a problem of the storage driver :/. You could give a try with ceph, you can set `export KUBEVIRT_STORAGE=rook-ceph-default`  before `cluster-up`

jean-edouard replied:
I need a storage class that supports FS + RWX. I'll see if I can make cephfs work.

jean-edouard replied:
With fsGroup=107 and rook-cephfs as a storage class:
```
/var/run/kubevirt-private/libvirt:
total 0
drwxr-sr-x. 3 root qemu system_u:object_r:container_file_t:s0:c547,c660 19 Feb 15 15:06 qemu

/var/run/kubevirt-private/libvirt/qemu:
total 0
drwxrwsr-x. 2 root qemu system_u:object_r:container_file_t:s0:c547,c660 0 Feb 15 15:06 swtpm
```
The `libvirt` directory is still a problem:
```
bash-5.1$ touch /var/run/kubevirt-private/libvirt/test
touch: cannot touch '/var/run/kubevirt-private/libvirt/test': Permission denied
```

alicefr replied:
mh and can you add also the `/var/run/kubevirt-private/libvirt` under the VolumeMount and then the other subdirectories? It's just a guess, not sure if it's possible

jean-edouard replied:
> mh and can you add also the `/var/run/kubevirt-private/libvirt` under the VolumeMount and then the other subdirectories? It's just a guess, not sure if it's possible

Do you mean adding all the intermediate directories as emptydirs? It might work, but that's not pretty!

alicefr replied:
Well, it is a way to say that you want also the upper directory writable. Otherwise, k8s/CSI might assume you don't need to write into it and set it readonly

jean-edouard replied:
Testing now. If it works, it sounds like you prefer that approach over the init container trick. @rmohr WDYT?

alicefr replied:
> Testing now. If it works, it sounds like you prefer that approach over the init container trick. @rmohr WDYT?

Yes, I'd definitely prefer this to the init container

jean-edouard replied:
Seems to work fine with cephfs, will test nfs after. However, it does require `fsGroup` to be set, as the swtpm directory doesn't allow writes from _other_.
@xpivarc @acardace from a PSA point of view, is it acceptable to set `fsGroup` to 107 under the pod security context? Thanks

xpivarc replied:
PSA should not be a problem here. But I remember we did not want to use fsGroup in the past. I can't remember why.

rmohr replied:
I think we were setting selinux contexts and permissions before from virt- handler. It is simple and reliable and works for all storage classes. Unless there are some side effects which I am not aware of which we want to avoid nowadays I would recommend virt-handler as the place to ensure that permissions are correct.

jean-edouard replied:
@rmohr do you have an example of that? I tried grepping around for it and couldn't find anything.
Also, implementing a hook in the virt-launcher code to wait for virt-handler to come and fix things doesn't really feel "cheaper" to me that a couple of emptyDirs, or even a small init container...

alicefr replied:
@jean-edouard did it work also for ceph-csi?
If you don't like listing too many directories with emptyDir, do you think it would be possible to create only the upper directory and then let virt-launcher create the subdirectories at start time?

rmohr replied:
@jean-edouard maybe I missed the part that this directory needs to be ready when libvirt starts already, and not when qemu starts? Is that the case?

jean-edouard replied:
> @jean-edouard maybe I missed the part that this directory needs to be ready when libvirt starts already, and not when qemu starts? Is that the case?

The directory needs to be ready even before libvirt starts, as virt-launcher first prepares qemu.conf in there.

jean-edouard replied:
> @jean-edouard did it work also for nfs-csi? If you don't like listing too many directories with emptyDir, do you think it would be possible to create only the upper directory and then let virt-launcher create the subdirectories at start time?

Yes it did work for nfs-csi. Every sub-directory that's not explicitly created via init-container/emptyDir will not be writable. In this case I'm pretty sure we use every sub-directory of `/var/run/kubevirt-private/libvirt/qemu/swtpm`.
However, I didn't have to do anything for `/var/run/kubevirt-private/var/lib/swtpm-localca` since nothing writes to `var/` or `lib/`.

rmohr replied:
>> @jean-edouard maybe I missed the part that this directory needs to be ready when libvirt starts already, and not when qemu starts? Is that the case?

> The directory needs to be ready even before libvirt starts, as virt-launcher first prepares qemu.conf in there.

If that is the case, then I am happy with any solution which does not use an additional init container. So sgtm. Thanks. :+1:

jean-edouard commented:
/retest

jean-edouard commented:
/retest

jean-edouard commented:
/hold cancel
The hold was there to wait for a swtpm update to remove the workaround.
swtpm has been updated to 0.8.0 and the workaround has been removed.

alicefr commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,75 @@
+package backendstorage
```
Could we move this to `pkg/storage` directory?

alicefr commented on pkg/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,75 @@
+		},
+	}
+
+	_, err = client.CoreV1().PersistentVolumeClaims(vmi.Namespace).Create(context.Background(), pvc, metav1.CreateOptions{})
```
This would fail also if the PVC already exist, do we want this?Or in this case we simply skip it?

jean-edouard replied:
The PVC shouldn't exist, since we tested for it earlier in the function, and it shouldn't be called multiple times in parallel.
I can still check for `errors.IsAlreadyExists(err)` and succeed in that case.

alicefr commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +96,126 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+			}, 300)).To(Succeed(), "the state of the TPM did not persist")
+		}
+
+		DescribeTable("[Serial]should persist TPM secrets across", func(op1, op2 string) {
```
Why do we need serial?

jean-edouard replied:
Because we modify the CR

vasiliy-ul commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1115,6 +1129,12 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
 		}
+
+		// ensure that the backend storage associated with the VMI is ready before creating the pod
+		if !backendStorageReady {
```
Currently, `backendstorage.CreateIfNeeded` always returns either `true, nil` or `false, err` (if `err != nil`). Is there really a case when `backendStorageReady` will result in `false` and we get with delaying pod creation? Would it make sense to simplify `CreateIfNeeded` and make it simply return an error?

vasiliy-ul commented on staging/src/kubevirt.io/api/core/v1/types.go:
```diff
@@ -2226,6 +2226,10 @@ type KubeVirtConfiguration struct {
 	SeccompConfiguration           *SeccompConfiguration             `json:"seccompConfiguration,omitempty"`
+
+	// VMStateStorageClass is the name of the storage class to use for the PVCs created to preserve VM state, like TPM.
+	// The storage class must support RWX in filesystem mode.
```
Are there requirements for the binding mode `WaitForFirstConsumer` vs. `Immediate`? Or it will work with any mode?

jean-edouard replied:
I think it should work in any mode, as we don't need to mess with the PVC prior to pod creation.

alicefr replied:
Agree with @jean-edouard, it shouldn't make any differences the PVC needs to be bound when we deploy the pod

vasiliy-ul commented:
Overall, looks good to me. There are just two discussions that are marked as resolved, but I do not see that the actual concerns have been addressed: https://github.com/kubevirt/kubevirt/pull/8156#discussion_r928507612 and https://github.com/kubevirt/kubevirt/pull/8156#discussion_r928147207. Maybe smth got lost during rebase?

jean-edouard commented:
> Overall, looks good to me. There are just two discussions that are marked as resolved, but I do not see that the actual concerns have been addressed: [#8156 (comment)](https://github.com/kubevirt/kubevirt/pull/8156#discussion_r928507612) and [#8156 (comment)](https://github.com/kubevirt/kubevirt/pull/8156#discussion_r928147207). Maybe smth got lost during rebase?

Indeed! Looks like I marked those as resolved when they clearly aren't, not sure what happened there... Thank you for the heads-up!

jean-edouard commented:
@vasiliy-ul both comments should now be truly resolved!

alicefr commented on pkg/storage/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,101 @@
+
+		return nil
+	}
+	if err != nil {
```
nit: you could simply return directly the error

alicefr commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,184 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+		DescribeTable("[Serial]should persist TPM secrets across", Serial, func(op1, op2 string) {
+			By("Setting the backend storage class to the default for RWX FS or skip if missing")
+			storageClass, exists := libstorage.GetRWXFileSystemStorageClass()
+			if !exists {
```
A bit pedantic, I know. But I was wondering if we should use labels as @xpivarc mentioned in this [PR](https://github.com/kubevirt/kubevirt/pull/9200#issuecomment-1431498870). Feel free to address this in a later PR

jean-edouard replied:
That's actually a really good point, skips are bad! Replacing the skip with an error, since it's a sig-storage test.

alicefr commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,184 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+
+			By("Waiting for the VM to start")
+			Eventually(func() error {
+				vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
```
We will probably need wait functions also for VirtualMachine types

alicefr commented:
Looks good to me!
/lgtm

vasiliy-ul commented on pkg/storage/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,101 @@
+		// The alternative is to remove this `if` block, allowing the PVC to persist after the VMI is deleted.
+		// However, that would pose security and littering concerns.
+		ownerReferences = []metav1.OwnerReference{{
+			APIVersion:         corev1.VirtualMachineInstanceGroupVersionKind.GroupVersion().String(),
+			Kind:               corev1.VirtualMachineInstanceGroupVersionKind.Kind,
+			Name:               vmi.Name,
+			UID:                vmi.UID,
+			Controller:         pointer.BoolPtr(true),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}}
```
Probably this can be simplified with `NewControllerRef` from `k8s.io/apimachinery/pkg/apis/meta/v1`:

```suggestion
			*metav1.NewControllerRef(vmi, v1.VirtualMachineInstanceGroupVersionKind),
		}}
```

vasiliy-ul commented on pkg/storage/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,101 @@
+	PVCSize   = "10Mi"
+)
+
+func PersistentPVCForVMI(vmi *corev1.VirtualMachineInstance) string {
```
nit: would it be better to name the function, e.g. `BackendStoragePVCforVMI`? Because PVC already stands for *persistent* volume claim. So we have 'persistent' twice here.

jean-edouard replied:
This is already the `backendstorage` package, so just `PVCForVMI()` should do.

vasiliy-ul commented on pkg/virt-controller/services/rendervolumes.go:
```diff
@@ -301,6 +303,61 @@ func withAccessCredentials(accessCredentials []v1.AccessCredential) VolumeRender
+	return func(renderer *VolumeRenderer) error {
+		if backendstorage.HasPersistentTPMDevice(vmi) {
+			volumeName := vmi.Name + "-tpm"
+			pvcName := backendstorage.PVCPrefix + vmi.Name
```
```suggestion
			pvcName := PersistentPVCForVMI(vmi)
```

vasiliy-ul commented on pkg/storage/backend-storage/backend-storage.go:
```diff
@@ -0,0 +1,101 @@
+		return nil
+	}
+
+	_, err := client.CoreV1().PersistentVolumeClaims(vmi.Namespace).Get(context.Background(), PVCPrefix+vmi.Name, metav1.GetOptions{})
```
```suggestion
	_, err := client.CoreV1().PersistentVolumeClaims(vmi.Namespace).Get(context.Background(), PersistentPVCForVMI(vmi), metav1.GetOptions{})
```

vasiliy-ul reviewed (commented):
/approve
/hold

Looks good! Added a hold in case you want to address the nits.

jean-edouard commented:
/hold cancel
Last comments should now be addressed.

jean-edouard commented:
/retest

vladikr commented:
/retest

poojaghumre commented:
Hi.. Can this feature get backported to the 0.59 patch release once merged?

jean-edouard commented:
> /hold Can we merge a proposal first?

Done, thanks @xpivarc !
/hold cancel

jean-edouard commented:
> /hold Can we merge a proposal first?

Done, thanks @xpivarc !
/hold cancel

xpivarc commented:
> Hi.. Can this feature get backported to the 0.59 patch release once merged?

Unfortunately I don't think this can be done with our backport policy.

rmohr commented:
Do snapshots have to consider this PVC as well?

jean-edouard commented:
> Do snapshots have to consider this PVC as well?

That's a good point, they probably do, I'll look into it, thanks!

jean-edouard commented:
> > Do snapshots have to consider this PVC as well?
> 
> That's a good point, they probably do, I'll look into it, thanks!

Added as a new commit.
It is a bit awkward since the only RWX FS storage class supported by kubevirtci, nfs-csi, doesn't support snapshots, therefore the functional test doesn't test much...
I'm not sure if cephfs comes with snaphot support, or if it can be configured for it, but that could be a solution to properly test the feature.
For now, if you'd rather block snapshots for VMs with persistent backend storage instead of merging a half-tested feature, let me know and I'll do that instead. Thanks!

jean-edouard commented:
> > > Do snapshots have to consider this PVC as well?
> > 
> > 
> > That's a good point, they probably do, I'll look into it, thanks!
> 
> Added as a new commit. It is a bit awkward since the only RWX FS storage class supported by kubevirtci, nfs-csi, doesn't support snapshots, therefore the functional test doesn't test much... I'm not sure if cephfs comes with snaphot support, or if it can be configured for it, but that could be a solution to properly test the feature. For now, if you'd rather block snapshots for VMs with persistent backend storage instead of merging a half-tested feature, let me know and I'll do that instead. Thanks!

So, cephfs does support snapshot, so I was able to test my rough implementation and it didn't quite work.
The PVC does get snapshotted and restored properly, however I didn't realize that the `restore` operation creates new PVCs and updates the name inside the VM object, as opposed to renaming/removing the existing PVCs to reuse their name.
This is a problem for backend storage, as the PVC is not a volume of the VM, therefore we have no way of telling the VM about the name of the restored PVC.

I have a couple ideas on how to fix that, but I definitely think it should be done as a separate effort.
So I removed the commit for snapshot support and replaced it with a change to the snapshot admitter to reject VMs that use backend storage for now.
I will be working on snapshot support next, but I think it would be nice to have basic persistent TPM supported as soon as possible.

@rmohr WDYT?

fossedihelm commented:
/cc

fossedihelm commented on pkg/virt-controller/services/rendervolumes.go:
```diff
@@ -330,6 +332,61 @@ func withAccessCredentials(accessCredentials []v1.AccessCredential) VolumeRender
+			if util.IsNonRootVMI(vmi) {
+				// For non-root VMIs, the TPM state lives under /var/run/kubevirt-private/libvirt/qemu/swtpm
+				// To persist it, we need the persistent PVC to be mounted under that location.
+				// /var/run/kubevirt-private is an emptyDir, and k8s wwould automatically create the right sub-directories under it.
```
mega nit(typo):
`wwould`

fossedihelm commented on pkg/virt-controller/services/rendervolumes.go:
```diff
@@ -330,6 +332,61 @@ func withAccessCredentials(accessCredentials []v1.AccessCredential) VolumeRender
 	}
 }
 
+func withTPM(vmi *v1.VirtualMachineInstance) VolumeRendererOption {
```
can we add a unit test for this?

fossedihelm commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,182 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+
+		DescribeTable("[Serial]should persist TPM secrets across", Serial, func(op1, op2 string) {
+			By("Setting the backend storage class to the default for RWX FS or skip if missing")
+			storageClass, exists := libstorage.GetRWXFileSystemStorageClass()
+			Expect(exists).To(BeTrue(), "No RWX FS storage class found")
```
There is an inconsistent behavior between this `Describe` and the subsequent `It`:
Here we will fail if the storageClass does not exist, later we silently Skip the test.
Maybe this can converge in the same behavior inside the `BeforeEach` block.
Thanks

fossedihelm replied:
In the `BeforeEach` you could also directly adjust the KubevirtCR (marking the whole `Context` as Serial), and the vmi itself

fossedihelm commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,182 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+				&expect.BExp{R: "MYSECRET"},
+			}, 300)).To(Succeed(), "failed to store secret into the TPM")
+
+			if op1 == "migrate" {
+				migrateVMI(vmi)
+			} else if op1 == "restart" {
+				restartVM(vm)
+			}
+
+			checkTPM(vmi)
+
+			if op2 == "migrate" {
+				migrateVMI(vmi)
+			} else if op2 == "restart" {
+				restartVM(vm)
+			}
+
+			checkTPM(vmi)
```
Instead of `op1` and `op2` you can switch using `ops ...string` and then loop through it with:
```suggestion
			for _, op := range ops{
				switch op {
				case "migrate":
					migrateVMI(vmi)
				case "restart":
					restartVM(vm)
				}
				checkTPM(vmi)
			}
```

fossedihelm commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,182 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+
+		migrateVMI := func(vmi *v1.VirtualMachineInstance) {
+			By("Migrating the VMI")
+			checks.SkipIfMigrationIsNotPossible()
```
this `Skip` can be done at the beginning of the test to save some time and resources

fossedihelm replied:
Got it! Thanks for the explanation :+1:

fossedihelm commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,182 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+	Context("with persistent TPM VM option enabled", func() {
+		restartVM := func(vm *v1.VirtualMachine) {
+			By("Stopping the VM")
+			err = virtClient.VirtualMachine(vm.Namespace).Stop(context.Background(), vm.Name, &v1.StopOptions{})
+			ExpectWithOffset(1, err).ToNot(HaveOccurred())
+			EventuallyWithOffset(1, func() error {
+				_, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+				return err
+			}, 300*time.Second, 1*time.Second).ShouldNot(Succeed())
+
+			By("Starting the VM")
+			err = virtClient.VirtualMachine(vm.Namespace).Start(context.Background(), vm.Name, &v1.StartOptions{})
```
Can I ask why didn't you use `.Restart()`?
Thanks

fossedihelm replied:
Agreed! In this case maybe you can rename the function to `stopAndStart` :slightly_smiling_face:

fossedihelm commented on tests/vmi_tpm_test.go:
```diff
@@ -85,3 +101,182 @@ var _ = Describe("[sig-compute]vTPM", decorators.SigCompute, func() {
+			err = virtClient.VirtualMachine(util.NamespaceTestDefault).Delete(context.Background(), vm.Name, &k8smetav1.DeleteOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Ensuring the PVC gets deleted")
+			Eventually(func() error {
+				_, err = virtClient.VirtualMachine(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+				if !errors.IsNotFound(err) {
+					return fmt.Errorf("VM %s not removed: %v", vm.Name, err)
+				}
+				_, err = virtClient.CoreV1().PersistentVolumeClaims(vm.Namespace).Get(context.Background(), backendstorage.PVCForVMI(vmi), k8smetav1.GetOptions{})
+				if !errors.IsNotFound(err) {
+					return fmt.Errorf("PVC %s not removed: %v", backendstorage.PVCForVMI(vmi), err)
+				}
+				return nil
+			}, 300*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
```
Probably you can avoid this check since the subsequent test will ensure the same

fossedihelm reviewed (commented):
Thanks @jean-edouard! It looks great overall. I added some comments below.
Let me know what do you think about them. Thanks :)

rmohr commented:
> I have a couple ideas on how to fix that, but I definitely think it should be done as a separate effort.
So I removed the commit for snapshot support and replaced it with a change to the snapshot admitter to reject VMs that use backend storage for now.
I will be working on snapshot support next, but I think it would be nice to have basic persistent TPM supported as soon as possible.

> @rmohr WDYT?

Yes, sounds good to me. I am just also not sure when looking into block volumes, and mayb having pre-populated PVCs if this feature is ready yet for general consumption? I have not plans right now to go deeper into reviewing your work here, so feel free to continue without my approval, but maybe move this for now behind a feature gate until we have more contrete plans on the missing gaps like snapshot/restore/block pvc/making pvc name discoverable/... ?

jean-edouard commented:
Thank you @fossedihelm for the great review!
I addressed most comments in the latest force-push.
I am now working on adding a unit test and a feature gate.

fossedihelm commented on tests/testsuite/kubevirtresource.go:
```diff
@@ -107,6 +107,7 @@ func AdjustKubeVirtResource() {
 		virtconfig.VMExportGate,
 		virtconfig.KubevirtSeccompProfile,
 		virtconfig.HotplugNetworkIfacesGate,
+		virtconfig.VMPersistentState,
```
Do we want to enable this fg for the whole test suite? Or do we want to restrict it to the only persistent vTPM tests?

jean-edouard commented:
/retest

vasiliy-ul commented:
Looks good.

/unhold
